File: RxTango/python/examples/multi_device_snapshot.py
"""Parallel snapshot of the same attribute from multiple devices.

Uses rx.from_iterable + flat_map to fire all reads concurrently, collecting
results into a list.  Errors from individual devices are replaced with None
via ops.catch so the snapshot continues even if a device is unreachable.

Mirrors `MultiDeviceSnapshot.java`.

Usage:
    python multi_device_snapshot.py [device1] [device2] ...

    Defaults to two reads of sys/tg_test/1 with different attributes
    (attribute names for a quick live test with a single device server).
"""

import asyncio
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent / "src"))

import reactivex as rx
import reactivex.operators as ops
from reactivex.scheduler.eventloop import AsyncIOScheduler

from rxtango import read_attribute


async def main() -> None:
    devices = sys.argv[1:] if len(sys.argv) > 1 else [
        "tango://localhost:10000/sys/tg_test/1",
        "tango://localhost:10000/sys/tg_test/1",
    ]
    attrs = ["double_scalar", "long_scalar"]

    loop      = asyncio.get_running_loop()
    scheduler = AsyncIOScheduler(loop)
    done      = asyncio.Event()
    results   = []

    pairs = list(zip(devices, attrs))
    print(f"Parallel snapshot of {len(pairs)} attribute(s):\n")

    rx.from_iterable(pairs).pipe(
        ops.flat_map(
            lambda da: read_attribute(da[0], da[1]).pipe(
                ops.map(lambda v: {"device": da[0], "attr": da[1], "value": v}),
                ops.catch(lambda e, _: rx.of(
                    {"device": da[0], "attr": da[1], "value": None, "error": str(e)}
                )),
            )
        ),
        ops.to_list(),
    ).subscribe(
        on_next=lambda items: [results.extend(items)],
        on_error=lambda e: (print(f"ERROR: {e}", file=sys.stderr), done.set()),
        on_completed=done.set,
        scheduler=scheduler,
    )

    await asyncio.wait_for(done.wait(), timeout=10.0)

    for item in results:
        if item.get("error"):
            print(f"  {item['device']}/{item['attr']}  ERROR: {item['error']}")
        else:
            print(f"  {item['device']}/{item['attr']}  = {item['value']}")


if __name__ == "__main__":
    asyncio.run(main())

File: RxTango/cpp/examples/multi_device_snapshot.cpp
/**
 * Multi-device parallel snapshot — concurrent reads from N devices, collected into a vector.
 *
 * Mirrors Python's multi_device_snapshot.py and Java's MultiDeviceSnapshot.java.
 *
 * All reads fire in parallel; per-device failures are isolated (on_error_return_item)
 * so one bad device never blocks the others.
 *
 * Usage:
 *   ./multi_device_snapshot [attr] [device1] [device2] ...
 *   defaults: double_scalar  sys/tg_test/1  sys/tg_test/1  (same device twice)
 */

#include <condition_variable>
#include <iostream>
#include <mutex>
#include <string>
#include <vector>

#include <rxcpp/rx.hpp>
#include <rxtango/rxtango.hpp>

int main(int argc, char* argv[]) {
    const std::string attr = argc > 1 ? argv[1] : "double_scalar";

    std::vector<std::string> devices;
    for (int i = 2; i < argc; ++i) devices.emplace_back(argv[i]);
    if (devices.empty()) {
        devices = { "tango://localhost:10000/sys/tg_test/1",
                    "tango://localhost:10000/sys/tg_test/1" };
    }

    std::cout << "Snapshot of " << devices.size() << " device(s), attribute: " << attr << "\n\n";

    std::mutex              m;
    std::condition_variable cv;
    bool                    done = false;

    // Mirrors: fromIterable → flat_map (concurrent) → on_error_return → to_list
    rxcpp::observable<>::iterate(devices)
        .flat_map([attr](const std::string& dev) {
            return rxtango::read_attribute<double>(dev, attr)
                .on_error_resume_next([dev](std::exception_ptr) {
                    // Isolate per-device failures with a sentinel value
                    return rxcpp::observable<>::just(std::numeric_limits<double>::quiet_NaN());
                });
        })
        .to_vector()
        .subscribe(
            [&devices, &attr](const std::vector<double>& results) {
                for (size_t i = 0; i < results.size(); ++i) {
                    std::cout << "  [" << i << "] " << devices[i] << "/" << attr
                              << " = " << results[i] << "\n";
                }
            },
            [&](std::exception_ptr e) {
                try { std::rethrow_exception(e); }
                catch (std::exception& ex) { std::cerr << "  ERROR: " << ex.what() << "\n"; }
                std::lock_guard<std::mutex> lk(m); done = true; cv.notify_one();
            },
            [&]() { std::lock_guard<std::mutex> lk(m); done = true; cv.notify_one(); }
        );

    std::unique_lock<std::mutex> lk(m);
    cv.wait(lk, [&]{ return done; });
    return 0;
}
